---
title: "NF Vocabulary / Schema"
author: "NF-OSI DCC"
date: "Last updated on `r Sys.Date()`"
output: 
  html_document:
    theme: lumen
    css: custom.css
    toc: true
    toc_float: 
      collapsed: false
    toc_depth: 4
params:
  schema_json: "../NF.jsonld"
  ext_classes_csv: "../ext_classes.csv"
  ext_relations_csv: "../ext_relations.csv"
  graph_view: false
  use_cache: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(data.table)
library(reactable)
library(htmltools)
# library(visNetwork)

source("graph.R")
source("docTemplate.R") 


basicTable <- function(dt, columns = c("Attribute", "Description")) {
  reactable(dt[, columns], 
          filterable = if(nrow(dt) > 20) TRUE else FALSE,
          pagination = FALSE,
          columns = list(
            Attribute = colDef(name = "Label", maxWidth = 250)
          ),
          wrap = FALSE,
          class = "term-table"
  )
}

expandedTable <- function(dt) {
  row_details <- function(index) {
    expanded <- dt[index, ]
    detail <- div(
      class = "detail",
      div(class = "detail-header",  expanded$Attribute, 
          span(class = "detail-title",  expanded$Description))
    )
    detail
  }
  
  

  reactable(dt[, c("Attribute", "Description")], 
          filterable = TRUE,
          pagination = FALSE,
          onClick = "expand",
          columns = list(
            Attribute = colDef(name = "Label", maxWidth = 250)
            # DependsOn = colDef(name = "Fields")
          ),
          details = row_details,
          wrap = FALSE,
          class = "term-table"
  )
}

templateTable <- function(dt) {
  # Reference range by linking to relevant class sections within document
  # or returning range values in plain text
  refRange <- function(value, index) {
    values <- strsplit(value, split = "|", fixed = TRUE)[[1]]
    if(any(grepl("#", values))) {
      htmltools::div(
        lapply(values, function(value) htmltools::p(
          htmltools::tags$a(href = value, sub("#", "", value)), " "))
        )
    } else {
      value
    }  
  }
  
  reactable(dt,
            columns = list(
              Field = colDef(minWidth = 180,
                             cell = function(value, index) { div(style = list(fontWeight = 600), value)}),
              Description = colDef(minWidth = 300),
              Required = colDef(maxWidth = 100),
              ValidRange = colDef(name = "Valid Range", cell = refRange)
            ),
            rowStyle = function(index) {
              if (dt[index, "Required"]) {
                list(background = "lemonchiffon")
                }
              },
            pagination = FALSE,
            wrap = TRUE,
            class = "term-table")
}


module_section <- function(module, 
                           submodule, 
                           submodule_id = submodule) {
  
  file <- glue::glue("../modules/{module}/{submodule}.yaml")
  data <- yaml::read_yaml(file)
  
  for(key in submodule_id) {
    data <- data$enums[[key]]$permissible_values
    
    cat(paste0("\n", "#### ", key, " {#", key, " .tabset .tabset-fade .tabset-pills}", "\n"))
    
    table <- rbindlist(data, fill = TRUE, idcol = "Attribute")
    setnames(table, "description", "Description", skip_absent = TRUE)
     if(!"Description" %in% names(table)) table[, Description := ""]
    
    table <- table[, .(Attribute, Description)]
  
    print(
      tagList(
        expandedTable(table), tags$br()))
  }
}


```

```{r process_schema_table, echo=FALSE}

schema <- jsonlite::read_json(params$schema_json)
schema <- schema$`@graph`

if(!params$use_cache) {
  templates <- c(# genomics/transcriptomics
                GenomicsAssayTemplate = "Genomics_Assay_Template", 
                WGSTemplate = "WGS_Template", 
                WESTemplate = "WES_Template",
                EpigeneticsAssayTemplate = "Epigenetics_Assay_Template",
                RNASeqTemplate = "RNASeq_Template",
                ScRNASeqTemplate = "ScRNASeq_Template",

                # genomics/transcriptomics -- processed
                ProcessedAlignedReadsTemplate = "Processed_Aligned_Reads_Template",
                ProcessedVariantCallsTemplate = "Processed_Variant_Calls_Template",
                ProcessedExpressionTemplate = "Processed_Expression_Template",
                
                # proteomics
                ProteomicsAssayTemplate = "Proteomics_Assay_Template", 
                
                # imaging
                ImagingAssayTemplate = "Imaging_Assay_Template", 
                MRIAssayTemplate = "MRI_Assay_Template",
                
                # other
                PharmacokineticsAssayTemplate = "Pharmacokinetics_Assay_Template", # drug assays type
                PlateBasedReporterAssayTemplate = "Plate_Based_Reporter_Assay_Template",
                ClinicalAssayTemplate = "Clinical_Assay_Template",
                # PatientTimepointsTemplate = "Patient_Timepoints_Template",
                ProtocolTemplate = "Protocol_Template",
                SourceCodeTemplate = "Source_Code_Template",
                LightScatteringAssayTemplate = "Light_Scattering_Assay_Template")
  
  docTemplate(templates, schema = schema)
}
  
# Splits terms into Property or Class using `Type`
# schema_class <- schema %>%
#   filter(Type == "Class")

# Not displayed yet
# schema_prop <- schema %>%
#  filter(Type == "Property")

```

## Annotation Templates

:::info

Annotation templates are spreadsheet templates that allow contributors to annotate resources uploaded to the repository.
See interactive use of these templates at our [NF Data Curator App](https://shiny.synapse.org/users/rallaway/NF_data_curator/).
Templates implement "minimum metadata" standards specific to the type of data/resource (hence variants exist for assay types and "raw" vs "processed" data).
Templates also contain common components, e.g. many will collect core sample info associated with the data.  

Unless the template field is free-text, it is meant to be filled by the contributor using the ontology terms/controlled vocabulary defined here.
For example, the "assay" property allows the contributor to use [terms under Assay](#assay).

:::

#### Genomics Assay Data Templates {#Genomics_Assay_Templates .tabset .tabset-fade .tabset-pills}

:::info

Raw or processed data files from a genomics assay should be annotated using one of these templates.  

:::

##### Genomics Assay (Generic)

```{r echo=FALSE }

templateTable(read.csv("templates/Genomics_Assay_Template.csv"))

```

##### Epigenetics Assay
```{r echo=FALSE }

templateTable(read.csv("templates/Epigenetics_Assay_Template.csv"))

```


##### WES Assay

```{r echo=FALSE }

templateTable(read.csv("templates/WES_Template.csv"))

```

##### WGS Assay

```{r echo=FALSE }

templateTable(read.csv("templates/WGS_Template.csv"))

```

##### Bulk RNA-Seq Assay

```{r echo=FALSE }

templateTable(read.csv("templates/RNASeq_Template.csv"))

```

##### Single-cell RNA-Seq Assay

```{r echo=FALSE }

templateTable(read.csv("templates/ScRNASeq_Template.csv"))

```


##### Processed Aligned Reads

```{r echo=FALSE }

templateTable(read.csv("templates/Processed_Aligned_Reads_Template.csv"))

```

##### Processed Variant Calls

```{r echo=FALSE }

templateTable(read.csv("templates/Processed_Variant_Calls_Template.csv"))

```

##### Processed Expression

```{r echo=FALSE }

templateTable(read.csv("templates/Processed_Expression_Template.csv"))

```

#### Proteomics Assay Templates {#Proteomics_Assay_Templates .tabset .tabset-fade .tabset-pills}

##### Proteomics Assay (Generic)
```{r echo=FALSE }

templateTable(read.csv("templates/Proteomics_Assay_Template.csv"))

```


#### Imaging Assay Templates {#Imaging_Assay_Templates .tabset .tabset-fade .tabset-pills}

##### Imaging Assay (Generic)
```{r echo=FALSE }

templateTable(read.csv("templates/Imaging_Assay_Template.csv"))

```

##### MRI Assay
```{r echo=FALSE }

templateTable(read.csv("templates/MRI_Assay_Template.csv"))

```

#### Other Assay Templates {#Other_Assay_Templates .tabset .tabset-fade .tabset-pills}

##### Pharmacokinetics Assay

```{r echo=FALSE }

templateTable(read.csv("templates/Pharmacokinetics_Assay_Template.csv"))

```

##### Plate Based Reporter Assay
```{r echo=FALSE }

templateTable(read.csv("templates/Plate_Based_Reporter_Assay_Template.csv"))

```

#### Clinical Assay Templates {#Clinical_Assay_Templates .tabset .tabset-fade .tabset-pills}

##### Clinical Assay (Generic)
```{r echo=FALSE }

templateTable(read.csv("templates/Clinical_Assay_Template.csv"))

```

#### Other Assays {.tabset .tabset-fade .tabset-pills}


##### Light Scattering
```{r echo=FALSE }

templateTable(read.csv("templates/Light_Scattering_Assay_Template.csv"))

```

#### Non-Assay Templates {#Non_Assay_Template .tabset .tabset-fade .tabset-pills}

##### Protocol
```{r echo=FALSE }

templateTable(read.csv("templates/Protocol_Template.csv"))

```

##### Source Code
```{r echo=FALSE }

templateTable(read.csv("templates/Source_Code_Template.csv"))

```


## Annotation Vocabulary

:::info

These are standard terms available to NF resource contributors for annotation of their resource. 
Terms are grouped into modules. 
When using these terms, contributors are helping to label and classify resources using standard semantics for improved interoperability and findability.

:::

### Assay 

```{r assay, echo=FALSE,results='asis'}

module_section("Assay", "Assay", "AssayEnum")

```

```{r platform, echo=FALSE,results='asis'}

module_section("Assay", "Platform", "PlatformEnum")

```

##### Relations Graph

:::info

This partial graph view logically relates **assays** to metadata **templates** available at the [NF Data Curator App](https://shiny.synapse.org/users/rallaway/NF_data_curator/).
For example, assays under the classification of `Imaging_Assay` currently uses a generic `Imaging_Assay_Template` for annotation.
More specialized templates may be made available as needed for specific assays.

:::

```{r schema_ext_assay_template, out.width="100%", echo=FALSE, eval=params$graph_view}

schema_ext <- readExtSchema(params$schema_csv, params$ext_classes_csv)
assay <- getNodesEdges(schema_ext, "Assay", "A", 
                       nodes = list(color = list(A = "plum", C = "indigo"), 
                                    font.color = list(A = "black", C = "white")))
template <- getNodesEdges(schema_ext, "Template", "T", use_id = T, 
                          nodes = list(color = list(A = "pink", C = "firebrick"), 
                                       font.color = list(A = "black", C = "white")))
g_assay_template <- c2Cluster(assay, template, "suggests", params$ext_relations_csv)
defaultGraph(g_assay_template)

```


### Information Entity

```{r resource, echo=FALSE, results='asis'}

module_section("Data", "Resource", "Resource")
```

```{r data, echo=FALSE, results='asis'}

module_section("Data", "Data", "Data")
```

```{r datasubtype, echo=FALSE, results='asis'}

module_section("Data", "Data", "DataSubtypeEnum")
```


```{r fileformat, echo=FALSE, results='asis'}

module_section("Data", "FileFormat", "FileFormatEnum")
```


### Sample

:::info

Sample typically refers to a **biosample**, but on some rare occasions can refer to an inorganic sample from which data are generated.
For biosamples, there is a distinction between individual-level and specimen-level data.
Data can be linked to individual-level sample information such as sex, species, diagnosis, and genotype.
Data can be linked to specimen-level information such as sample site (the organ or body part), specimen type or cell type, tumor class (if specimen is a tumor), and specimen state.

:::


```{r sample_specimenType_tissue, echo=FALSE, results='asis'}

module_section("Sample", "SpecimenType", "Tissue")

```

```{r sample_specimenType_other, echo=FALSE, results='asis'}

module_section("Sample", "SpecimenType", "OrganismSubstance")

```


```{r sample_tumor, echo=FALSE, results='asis'}

module_section("Sample", "Tumor", "Tumor")

```

```{r sample_sex, echo=FALSE, results='asis'}

module_section("Sample", "Sex", "SexEnum")

```

```{r sample_diagonosis, echo=FALSE, results='asis'}

module_section("Sample", "Diagnosis", "DiagnosisEnum")

```


```{r sample_genotype, echo=FALSE, results='asis'}

module_section("Sample", "Genotype", "Genotype")

```


```{r sample_species, echo=FALSE, results='asis'}

module_section("Sample", "Species", "SpeciesEnum")

```

```{r sample_cell, echo=FALSE, results='asis'}

module_section("Sample", "Cell", "Cell")

```

```{r sample_cellmodel, echo=FALSE, results='asis'}

module_section("Sample", "CellLineModel", "CellLineModel")

```


```{r sample_mousemodel, echo=FALSE, results='asis'}

module_section("Sample", "MouseModel", "MouseModel")

```


### Experiment

:::info

Terminology to help characterize experiment.

:::

```{r biosample_module, echo=FALSE, results='asis'}

module_section("Experiment", "Unit", "TimeUnit")

```


## Other

#### Reserved Properties {#reserved .tabset .tabset-fade .tabset-pills}

:::info

Reserved annotation properties also describe the resource in some way, but they are not expected to be edited directly by contributors. 
They are computed/automated annotations set by the data platform and DCC activities.

:::

##### Terms
```{r reserved_properties_table, echo=FALSE,eval=F}

reserved_properties_table <- schema %>%
  filter(SubOf %in% c("dccProperty", "synapseProperty")) %>%
  select(Attribute, Description, DependsOn)

basicTable(reserved_properties_table)
```


