---
title: "NF Vocabulary / Schema"
author: "NF-OSI DCC"
date: "Last update on `r Sys.Date()`"
output: 
  html_document:
    theme: lumen
    css: custom.css
    toc: true
    toc_float: 
      collapsed: false
    toc_depth: 4
params:
  schema_csv: "../NF.csv"
  ext_classes_csv: "../ext_classes.csv"
  ext_relations_csv: "../ext_relations.csv"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(reactable)
library(htmltools)


source("graph.R")
source("docTemplate.R")

basicTable <- function(dt, columns = c("Attribute", "Description")) {
  reactable(dt[, columns], 
          filterable = if(nrow(dt) > 20) TRUE else FALSE,
          pagination = FALSE,
          columns = list(
            Attribute = colDef(name = "Label", maxWidth = 250)
          ),
          wrap = FALSE,
          class = "term-table"
  )
}

expandedTable <- function(dt) {
  row_details <- function(index) {
    expanded <- dt[index, ]
    detail <- div(
      class = "detail",
      div(class = "detail-header",  expanded$Attribute, 
          span(class = "detail-title",  expanded$Description))
    )
    detail
  }

  reactable(dt[, c("Attribute", "Description")], 
          filterable = TRUE,
          pagination = FALSE,
          onClick = "expand",
          columns = list(
            Attribute = colDef(name = "Label", maxWidth = 250)
            # DependsOn = colDef(name = "Fields")
          ),
          details = row_details,
          wrap = FALSE,
          class = "term-table"
  )
}

templateTable <- function(dt) {
  # Reference range by linking to relevant class sections within document
  # or returning range values in plain text
  refRange <- function(value, index) {
    if(grepl("#", value)) {
      htmltools::tags$a(href = value, sub("#", "", value))
    } else {
      value
    }  
  }
  
  reactable(dt,
            columns = list(
              Field = colDef(minWidth = 180),
              Description = colDef(minWidth = 300),
              Required = colDef(maxWidth = 100),
              ControlledVocab = colDef(name = "Controlled\nVocab",
                                       cell = refRange)
            ),
            pagination = FALSE,
            wrap = TRUE,
            class = "term-table")
}
```

```{r process_schema_table, echo=FALSE}

# To avoid overwhelming info and keep page neat, process only selected columns
# Aside from default schematic columns, 
# we require internal columns prefixed with `.` to have a useful table
schema <- read.csv(params$schema_csv) %>%
  select(Attribute, Description, Required, Valid.Values, DependsOn, Parent, 
         ID = .ID, Type = .Type, Root = .Root, SubclassOf = .SubclassOf,
         Range = .Range, EditorNote = .EditorNote)


# docTemplate needs the unmodified schema table
docTemplate(schema)

# Splits terms into Property or Class using `Type`
# schema_class <- schema %>%
#   filter(Type == "Class")

# Not displayed yet
# schema_prop <- schema %>%
#  filter(Type == "Property")

```

## Annotation Classes

:::info

These are standard terms available to NF resource contributors for annotation of their resource.
When contributors use these terms, they are helping to label and classify their resource for improved interoperability and findability.

:::

### Assay Module

#### Assay {#Assay .tabset .tabset-fade .tabset-pills}

##### Terms

```{r assays_table, echo=FALSE}

assays_table <- schema %>%
  filter(Root == "Assay") %>%
  select(Attribute, Description, DependsOn)

expandedTable(assays_table)

```


##### Relations Graph

:::info

This partial graph view logically relates **assays** to metadata **templates** available at the [NF Data Curator App](https://shiny.synapse.org/users/rallaway/NF_data_curator/).
For example, assays under the classification of `Imaging_Assay` currently uses a generic `Imaging_Assay_Template` for annotation.
More specialized templates may be made available as needed for specific assays.

:::

```{r schema_ext_assay_template, out.width="100%", echo=FALSE}

schema_ext <- readExtSchema(params$schema_csv, params$ext_classes_csv)
assay <- getNodesEdges(schema_ext, "Assay", "A", 
                       nodes = list(color = list(A = "plum", C = "indigo"), 
                                    font.color = list(A = "black", C = "white")))
template <- getNodesEdges(schema_ext, "Template", "T", use_id = T, 
                          nodes = list(color = list(A = "pink", C = "firebrick"), 
                                       font.color = list(A = "black", C = "white")))
g_assay_template <- c2Cluster(assay, template, "suggests", params$ext_relations_csv)
defaultGraph(g_assay_template)

```

#### Platform {#Platform .tabset .tabset-fade .tabset-pills}

##### Terms

```{r platforms_table, echo=FALSE}

platforms_table <- schema %>%
  filter(Parent == "platform") %>%
  select(Attribute, Description, DependsOn)

basicTable(platforms_table)
```


##### Relations Graph

:::info

This partial graph view logically relates **assays** to common **platforms**.

:::

*Documentation currently in development.*

#### Parameters

:::info

The parameters used should be specified specific to the assay and platform.

:::

##### NGS Parameter {#NGS_Parameter}

###### Library Prep {#Library_Prep}

```{r echo=FALSE}

prep_params <- schema %>%
  filter(Parent == "libraryPrep") %>%
  select(Attribute, Description, DependsOn)

basicTable(prep_params)
```

###### Library Prep Method {#Library_Preparation_Method}

```{r echo=FALSE}

prep_method_params <- schema %>%
  filter(Parent == "libraryPrepMethod") %>%
  select(Attribute, Description, DependsOn)

basicTable(prep_method_params)
```


###### Read Pair {#Read_Pair}

```{r echo=FALSE}

read_params <- schema %>%
  filter(Parent == "readPair") %>%
  select(Attribute, Description, DependsOn)

basicTable(read_params)
```

###### Read Pair Orientation {#Read_Pair_Orientation}

```{r echo=FALSE}

orientation_params <- schema %>%
  filter(Parent == "readPairOrientation") %>%
  select(Attribute, Description, DependsOn)

basicTable(orientation_params)
```

###### Run Type {#Run_Type}
```{r echo=FALSE}

run_params <- schema %>%
  filter(Parent == "runType") %>%
  select(Attribute, Description, DependsOn)

basicTable(run_params)
```



### Data Module

#### Data Type {#Data .tabset .tabset-fade .tabset-pills}

##### Terms

```{r data_types_table, echo=FALSE}

data_types_table <- schema %>%
  filter(Parent == "dataType") %>%
  select(Attribute, Description, DependsOn)

basicTable(data_types_table)
```

##### Relations Graph

:::info

This partial graph view logically relates **data types** to **assays**.

:::

*Documentation currently in development.*

#### File Format {#File_Format .tabset .tabset-fade .tabset-pills}

##### Terms

:::info

Certain formats are emphasized as <div class='warning badge' display=inline>proprietary</div>, which make data less interoperable/reusable, as opposed to our <div class='good badge' display=inline>preferred</div> (open) formats.

:::

```{r data_formats_table, echo=FALSE}

data_formats_table <- schema %>%
  filter(Parent == "fileFormat") %>%
  select(Attribute, Description, Note = SubclassOf)
  
# A little more complicated than the other tables
reactable(data_formats_table, 
          filterable = TRUE,
          pagination = FALSE,
          columns = list(
            Attribute = colDef(name = "Label", maxWidth = 150),
            Note = colDef(
              maxWidth = 100,
              cell = function(value) {
                label <- switch(value, Preferred_Open_Format = "preferred", Proprietary_Format = "proprietary", "")
                class <- switch(value, Preferred_Open_Format = "good", Proprietary_Format = "warning", "")
                div(class = paste(class, "badge"), label)
              }
            )
          ),
          wrap = FALSE,
          class = "term-table"
  )

```

##### Relations Graph

:::info

This partial graph view logically organizes **data formats** and notes which formats are interconvertible.

:::

*Documentation currently in development.*

### Biosample Module

:::info

Biosample combines metadata at the individual-level and specimen-level.
Data can be linked to individual-level sample information such as sex, species, diagnosis, and genotype.

:::

<!-- Sample Individual --> 

#### Sex {#Sex .tabset .tabset-fade .tabset-pills}

##### Terms
```{r sex_table, echo=FALSE}

sex_table <- schema %>%
  filter(Parent == "sex") %>%
  select(Attribute, Description, DependsOn)

basicTable(sex_table)
```

#### Species {#Species .tabset .tabset-fade .tabset-pills}

##### Terms
```{r species_table, echo=FALSE}

species_table <- schema %>%
  filter(Parent == "species") %>%
  select(Attribute, Description, DependsOn)

basicTable(species_table)
```

#### Diagnosis {#Diagnosis .tabset .tabset-fade .tabset-pills}

##### Terms
```{r diagnosis_table, echo=FALSE}

diagnosis_table <- schema %>%
  filter(Parent == "diagnosis") %>%
  select(Attribute, Description, DependsOn)

basicTable(diagnosis_table)
```


#### Genotype {#Genotype .tabset .tabset-fade .tabset-pills}

##### Terms
```{r genotypes_table, echo=FALSE}

genotypes_table <- schema %>%
  filter(Root == "Genotype") %>%
  select(Attribute, Description, DependsOn)

basicTable(genotypes_table)
```


:::info

Data can be linked to specimen-level information such as sample site (the organ or body part), specimen tissue or cell type, tumor class (if specimen is a tumor), and specimen state.

:::

<!-- Sample Specimen --> 

#### Sample Site {.tabset .tabset-fade .tabset-pills}

##### Terms

```{r sample_site_table, echo=FALSE}

sample_site_table <- schema %>%
  filter(Parent %in% c("organ", "bodyPart")) %>%
  select(Attribute, Description, DependsOn)

basicTable(sample_site_table)
```


#### Sample Specimen

##### Tissue {#Tissue} 
```{r tissue_table, echo=FALSE}

tissue <- schema %>%
  filter(Parent %in% c("tissue", "cellType")) %>%
  select(Attribute, Description, DependsOn)

basicTable(tissue)
```

##### Cell / Cell Line {#Cell} 
```{r echo=FALSE}

cell_table <- schema %>%
  filter(SubclassOf %in% c("Primary_Cell", "Cell_Line_Model")) %>%
  select(Attribute, Description, DependsOn)

basicTable(cell_table)
```

#### Specimen Dissociation {#Dissociation_Method}

*Documentation in development.*

#### Specimen State {#Specimen_State}

*Documentation in development.*

#### Tumor Class {#Tumor .tabset .tabset-fade .tabset-pills}

##### Terms
```{r tumor_table, echo=FALSE}

tumor_table <- schema %>%
  filter(Parent == "tumorType") %>%
  select(Attribute, Description, DependsOn)

basicTable(tumor_table)
```

### Experiment Module

#### Gene Perturbation

##### Gene Perturbation Type {#Gene_Perturbation}
```{r echo=FALSE}

perturb_type <- schema %>%
  filter(Parent == "genePerturbationType") %>%
  select(Attribute, Description, DependsOn)

basicTable(perturb_type)
```

##### Gene Perturbation Technology {#Gene_Perturbation_Technology}

```{r echo=FALSE}

perturb_tech <- schema %>%
  filter(Parent == "genePerturbationTechnology") %>%
  select(Attribute, Description, DependsOn)

basicTable(perturb_tech)
```

### Cross-ref Module

:::info

This is a provisional module for annotations that help cross-reference data for internal tracking and inter-database linkages.

:::

#### Model System {#Model_System}

*Documentation in development.*

## Annotation Templates

:::info

Annotation templates are spreadsheet templates that allow contributors to annotate resources uploaded to the repository.
They are generated at the [NF Data Curator App](https://shiny.synapse.org/users/rallaway/NF_data_curator/).
Templates aim to implement "minimum metadata" standards specific to the type of data/resource.
Unless the template field is free-text, it is meant to be filled by the contributor using the ontology terms/controlled vocabulary defined here.
For example, the "assay" property allows the contributor to use [terms under Assay](#assay).

:::

#### Templates for Assay Data {#Assay_Template .tabset .tabset-fade .tabset-pills}

```{r standard_properties_table, echo=FALSE, eval=FALSE}
# NOT RUN
# Custom code to get unique set of properties used across all templates
standard_properties_table <- schema %>%
  filter(Parent %in% c("template")) %>%
  select(Attribute, Description, DependsOn)

basicTable(standard_properties_table)
```

##### Genomics Assay

```{r echo=FALSE }

templateTable(read.csv("templates/Genomics_Assay_Template.csv"))

```

##### Proteomics Assay
```{r echo=FALSE }

templateTable(read.csv("templates/Proteomics_Assay_Template.csv"))

```

##### Epigenetics Assay
```{r echo=FALSE }

templateTable(read.csv("templates/Epigenetics_Assay_Template.csv"))

```

##### Imaging Assay
```{r echo=FALSE }

templateTable(read.csv("templates/Imaging_Assay_Template.csv"))

```

##### MRI Assay
```{r echo=FALSE }

templateTable(read.csv("templates/MRI_Assay_Template.csv"))

```


##### Pharmacokinetics Assay

```{r echo=FALSE }

templateTable(read.csv("templates/Pharmacokinetics_Assay_Template.csv"))

```

##### Plate Based Reporter Assay
```{r echo=FALSE }

templateTable(read.csv("templates/Plate_Based_Reporter_Assay_Template.csv"))

```

#### Other Templates {#Non_Assay_Template .tabset .tabset-fade .tabset-pills}

##### Patient Timepoints
```{r echo=FALSE }

templateTable(read.csv("templates/Patient_Timepoints_Template.csv"))

```

## Miscellaneous

#### System Properties {#reserved .tabset .tabset-fade .tabset-pills}

:::info

System annotation properties also describe the resource in some way, but they are not expected to be edited directly by contributors. 
They are computed/automated annotations set by the data platform and DCC activities.

:::

##### Terms
```{r reserved_properties_table, echo=FALSE}

reserved_properties_table <- schema %>%
  filter(Root %in% c("dccProperty", "synapseProperty")) %>%
  select(Attribute, Description, DependsOn)

basicTable(reserved_properties_table)
```


